### DISCUSSION TITLE
Running TypeScript code on Miden

### DISCUSSION THREAD

**Discussion by cf** [0xMiden/miden-vm#1077] [Show and tell]
Hey guys just wanted to thank the contributors of the project for all the hard work, and share our progress in supporting Miden VM as target for our TypeScript compiler.

You can see an example of us compiling and running a TypeScript matrix math library in the Polygon browser sandbox:
https://twitter.com/QEDProtocol/status/1704853277018665399

Compiled the [poseidon2](https://www.npmjs.com/package/poseidon2) npm package to miden:
https://twitter.com/cmpeq/status/1705070834036142442

Right now lots of room for optimization and better function isolation (right now using mem_read/mem_write instead of proc scoped memory, better stack optimization, etc.)

Planning to open source once we fix the perf issues, but If anyone has a usecase they would like to test early give us a shout!

**Comment by bobbinth** [comment]
This is super cool! Thank you for the great work! Looking forward to playing around with this is publicly available.

One thing I'm curious about: what things did you find difficult to work with/understand? Basically, how can we improve both the VM and the docs.

**Reply by cf** [reply]
Everything was very well documented, and I noticed you guys added a miden IR repo (maybe I just missed this before haha), so we I'll definitely be switching over to that rather than just emitting raw miden assembly.

One thing that would be very nice would be official support for poseidon/poseidon2. I saw that you guys wrote a [benchmark](https://github.com/0xPolygonMiden/crypto/tree/main/benches#comparison) for Poseidon (so not sure if this is in the roadmap already.

**Reply by cf** [reply]
Also another tidbit, our team will be open sourcing our plonky2 winterfell verifier gadget in the coming weeks (currently only supports poseidon + poseidon2), so will definitely add a miden sample if you guys end up supporting poseidon!
